server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -1508,7 +1508,7 @@ public static DataLifecycle resolveLifecycle(ComposableIndexTemplate template, M
     public static DataLifecycle composeDataLifecycles(List<DataLifecycle> lifecycles) {
         DataLifecycle.Builder builder = null;
         for (DataLifecycle current : lifecycles) {
-            if (current.equals(Template.NO_LIFECYCLE)) {
+            if (current == Template.NO_LIFECYCLE) {
```

Comment:
Should we add a test that fails if `equals` is used here?

Reply:
Do you mean to throw an exception is someone is using `equals`? Yes, why not :)

Reply:
Or maybe better just implement equals as follows:
```
        @Override
        public boolean equals(Object o) {
            return this == o;
        }
```

Reply:
Ok, I am back, because both `equals` and `==` behave the same in this case, I cannot really make a unit test fail, unless you had in mind to mock `DataLifecycle` to ensure that equals is not called, but I am not sure we should go that far. If I am not mistaken, it's nicer to have `==` but since both are correct, I don't think we should go to such lengths. What do you think?

PS: the `equals` fails because they are not the same class type.